oauthproxy.go

```diff
@@ -579,16 +581,43 @@ func (p *OAuthProxy) GetRedirect(req *http.Request) (redirect string, err error)
 		redirect = req.Form.Get("rd")
 	}
 	if !p.IsValidRedirect(redirect) {
-		// Use RequestURI to preserve ?query
-		redirect = req.URL.RequestURI()
-		if strings.HasPrefix(redirect, p.ProxyPrefix) {
-			redirect = "/"
+		// Quirk: On reverse proxies that doesn't have support for
+		// "X-Auth-Request-Redirect" header or dynamic header/query string
+		// manipulation (like Traefik v1 and v2), we can try if the header
+		// X-Forwarded-Host exists or not.
+		if isForwardedRequest(req) && p.ReverseProxy {
```

Comment:
So I think we want this dependency on `ReverseProxy` to be viable and consumed in all uses of the `util.GetRequest*` helper methods that potentially pull from `X-Forwarded-*` headers -- so their output is consistent across all uses (cookies, redirect, logging, etc).

To make that convenient and have those functions still only take in a `*http.Request` argument wherever they are called -- maybe we need to inject the `ReverseProxy` state of the request early in its lifecycle into the requests's internal `context`? Not sure all the places these are called already have access to a `ctx.Context` argument to make it easy to add as a second parameter.

@JoelSpeed Is this scope a little bigger than this PR? Should we tackle expanding ReverseProxy for this (and maybe `X-Auth-Request-Redirect` as well) in its own PR?

Reply:
> @JoelSpeed Is this scope a little bigger than this PR? Should we tackle expanding ReverseProxy for this (and maybe X-Auth-Request-Redirect as well) in its own PR?

Let's get this ready to merge then and tackle this at a later point, we should make sure that's within the same release though otherwise we will break the functionality when we do make this change